consensus/istanbul/backend/announce.go

```diff
@@ -393,17 +402,40 @@ func (sb *Backend) generateAndGossipQueryEnode(version uint, enforceRetryBackoff
 		return err
 	}
 
-	queryEnodeDestAddresses := make([]common.Address, 0)
-	queryEnodePublicKeys := make([]*ecdsa.PublicKey, 0)
+	queryEnodeEncryptedEnodeURLParams := make([]*genEncryptedEnodeURLParam, 0)
+
+	var valProxyAssignments map[common.Address]*enode.Node
+	var selfEnodeURL string
+	if sb.IsProxiedValidator() {
+		valProxyAssignments, err = sb.proxyEngine.GetValidatorProxyAssignments()
```

Comment:
I found this implementation a bit hard to follow on first read. I would suggest refactoring this to have `GetValidatorProxyAssignments`, when this node is a proxy, return a map with only the validator it is assigned to peer with as keys, and only itself as values. That factoring would allow this function to have one code execution path for both validators and proxies, with the additional benefit that the proxy code could include logic in the `GetValidatorProxyAssignments` function to ensure that it does not query for validators that it is not assigned to, even if they are in the nodes enode table. (I am a strong advocate for the idea that proxies should have a full enode table at all times)

Reply:
As I mentioned before, I agree that having proxies be more independent will make the system more resilient.  However, doing that will be a major re-architecture change, and I think releasing a first version with where the validator is the main driver of the announce protocol is still very valuable to the end user.  I'm def. happy to put independent proxies as a roadmap item for the proxies, in general, but that will require a lot of design thought.

Reply:
I didn't mean to start a discussion about the relationship between proxies and validators. This comment is simply a suggested refactor that would eliminate some code, and flatten the control flow, around the call sites for `GetValidatorProxyAssignments`.